lib/Ceilometer/Client.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators       #-}
+
+--
+-- Copyright © 2013-2015 Anchor Systems, Pty Ltd and Others
+--
+-- The code in this file, and the program it is a part of, is
+-- made available to you by its authors as open source software:
+-- you can redistribute it and/or modify it under the terms of
+-- the 3-clause BSD licence.
+--
+--
+-- This module defines a convienient interface for clients
+-- of Ceilometer.
+--
+-- For flexibility use the Collector and Fold modules.
+--
+module Ceilometer.Client
+  ( -- * Combo
+    decodeAndFold
+    -- * Decoding
+  , decode, decodeWith
+    -- * Folding
+  , foldDecoded, foldDecodedWith
+    -- * Re-exports
+  , module C
+  ) where
+
+import           Control.Applicative
+import           Control.Lens
+import           Control.Monad
+import           Data.Maybe
+import qualified Data.Traversable    as T
+import           Data.Typeable
+import           Data.Word
+import           Pipes
+
+import           Ceilometer.Fold     as C
+import           Ceilometer.Infer    as C
+import           Ceilometer.Types    as C
+import           Vaultaire.Types
+
+
+-- | Decode and fold a stream of @SimplePoint@.
+--
+--   If the assertion that these points should be type `a` is wrong: run-time error.
+--
+--   If the assertion is true but the @SourceDict@ doesn't match what we expect
+--   for the fold: `Nothing`.
+--
+--   Otherwise we get a `FoldResult` matching that type.
+--
+--   Example usage: `decodeAndFold (undefined :: proxy PDCPU) env points`
```

Comment:
unless we have Signatures Sections `` foo (:: MyType) x y...``, either this or ``Data.Proxy`` doesn't really make a difference.